src/common/sstring.hh

```diff
@@ -647,6 +650,60 @@ inline string_type to_sstring(T value) {
     return sstring::to_sstring<string_type>(value);
 }
 
+
+// encode/decode
+template <typename Char, typename Size, Size Max>
+struct denc_traits<basic_sstring<Char, Size, Max>> {
+private:
+  using value_type = basic_sstring<Char, Size, Max>;
+public:
+  static constexpr bool supported = true;
+  static constexpr bool featured = false;
+  static constexpr bool bounded = false;
+
+  static void bound_encode(const value_type& s, size_t& p, uint64_t f=0) {
+    p += sizeof(Size) + s.size();
+  }
+
+  static void encode_nohead(const value_type& s,
+                            buffer::list::contiguous_appender& p)
+  {
+    auto len = s.size();
+    if (len) {
+      p.append(reinterpret_cast<const char*>(s.c_str()), len);
+    }
+  }
+
+  static void decode_nohead(size_t len, value_type& s,
+                            buffer::ptr::iterator& p)
+  {
+    s.reset();
+    if (len) {
+      s.append(reinterpret_cast<const Char*>(p.get_pos_add(len)), len);
```

Comment:
yes, if we use the old-style `decode()`, we will be using the wrapper of `denc()` : https://github.com/ceph/ceph/blob/master/src/include/denc.h#L1432. `copy_shallow()` tries to do the shallow copy but if what we want exceeds the segment's length, it will do in the expensive way.